### PR TITLE
New repo-updater scheduler

### PR DESCRIPTION
**Pull request by nicksnyder**
This PR implements a new scheduling algorithm for repo updater. The existing logic in reposlist.go is complicated and hard to follow. The proposed code in this PR is about 2/3s the size of the previous scheduler and I think is much easier to follow since it splits the system into two parts: a scheduler and an updater.

I did not make any attempt to solve other known problems (i.e. supporting repo renames), but I don't think I made any of these problems worse.

GitHub helpfully collapses the most interesting file in this diff: scheduler.go.

This is feature flagged and off by default, so I would like to merge this as-is.

I will open a followup PR to address observability.

### CODE REVIEW

File: cmd/repo-updater/internal/pkg/scheduler/scheduler.go [hunk 1]
```diff
@@ -0,0 +1,422 @@
...
+	})
+
+	for {
+		select {
+		case <-r.queue.notifyEnqueue:
+		case <-ctx.Done():
+			return
+		}
+
+		for repo := r.queue.dequeue(); repo != nil; {
```
Comment (nicksnyder) [inline cmd/repo-updater/internal/pkg/scheduler/scheduler.go]: TODO: need to add an extra state to keep track of what is being cloned

File: cmd/repo-updater/internal/pkg/scheduler/scheduler.go [hunk 2]
```diff
@@ -0,0 +1,422 @@
...
+	}
+
+	if len(q.background) > 0 {
+		front := q.background[0]
+		q.background = q.background[1:]
+		delete(q.backgroundIndex, front.URI)
+		return front
+	}
+	return nil
+}
```
[the remark this answers is not in the document]
Comment (nicksnyder) [reply cmd/repo-updater/internal/pkg/scheduler/scheduler.go]: The code should handle this correctly now (I ended up using a heap for this reason and because it helped solve the stampede problem too).

File: cmd/repo-updater/internal/pkg/scheduler/scheduler.go [hunk 3]
```diff
@@ -0,0 +1,422 @@
...
+	URI     api.RepoURI
+	URL     string
+	Enabled bool
+}
+
+// SourceRepoList is the set of repositories associated with a specific configuration source.
+type SourceRepoList map[api.RepoURI]*ConfiguredRepo
+
+// Repos holds all of the state.
+var Repos = repoList{
```
[the remark this answers is not in the document]
Comment (nicksnyder) [reply cmd/repo-updater/internal/pkg/scheduler/scheduler.go]: I figured out some names I am happy with.

File: cmd/gitserver/server/server.go [hunk 4]
```diff
@@ -205,6 +205,8 @@ func (s *Server) Handler() http.Handler {
 	// GitMaxConcurrentClones controls the maximum number of clones that
 	// can happen at once. Used to prevent throttle limits from a code
 	// host. Defaults to 5.
+	// TODO(nick): The limiting here is deprecated in favor of enforcing
+	// the limit globally in repo-updater (instead of per gitserver instance here).
```
Comment (nicksnyder) [inline cmd/gitserver/server/server.go]: TODO: disable these deprecated code paths when the new feature flag is enabled

Comment (nicksnyder) [reply cmd/gitserver/server/server.go]: Doesn't make sense to deprecate this code path since `ensureRevision` code path still exists.

**Comment by nicksnyder** [comment]
This is ready for review @keegancsmith @beyang . I would like to get this merged so we can start observing the behavior on dogfood (feature flag still defaults to off).

I have added a 💩ton of tests (totally worth it; found some things that would have been painful to debug otherwise).

I will open a followup PR regarding observability (this diff is getting too large already)

File: cmd/repo-updater/repos/scheduler.go [hunk 5]
```diff
@@ -0,0 +1,502 @@
...
+var requestRepoUpdate = func(ctx context.Context, repo *configuredRepo, since time.Duration) (*gitserverprotocol.RepoUpdateResponse, error) {
+	return gitserver.DefaultClient.RequestRepoUpdate(ctx, gitserver.Repo{Name: repo.name, URL: repo.url}, since)
+}
+
+// configuredLimiter returns a mutable limiter that is
+// configured with the maximum number of concurrent update
+// requests that repo-updater should send to gitserver.
+var configuredLimiter = func() *mutablelimiter.Limiter {
+	limiter := mutablelimiter.New(1)
+	conf.Watch(func() {
```
[the remark this answers is not in the document]
Comment (nicksnyder) [reply cmd/repo-updater/repos/scheduler.go]: I think the question is whether or not we want to block waiting for the real concurrency limit before starting to clone anything. I don't think it matters much either way but I leaned toward blocking.

[the remark this answers is not in the document]
Comment (keegancsmith) [reply cmd/repo-updater/repos/scheduler.go]: Well we start with the smallest limit, and it being smaller than the true limit isn’t bad.
